# How do I filter rows of a pandas DataFrame by column value ? 

🐼 Tuto on pandas by Data School - Exercice performed by Dorian.H Mekni 🥇 | Sun 29 Nov 2020

In [1]:
import pandas as pd 

In [2]:
movies = pd.read_csv('http://bit.ly/imdbratings')

In [3]:
movies.head()

star_rating                     title content_rating   genre  duration  \
0          9.3  The Shawshank Redemption              R   Crime       142   
1          9.2             The Godfather              R   Crime       175   
2          9.1    The Godfather: Part II              R   Crime       200   
3          9.0           The Dark Knight          PG-13  Action       152   
4          8.9              Pulp Fiction              R   Crime       154   

                                         actors_list  
0  [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1    [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2  [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
3  [u'Christian Bale', u'Heath Ledger', u'Aaron E...  
4  [u'John Travolta', u'Uma Thurman', u'Samuel L....

In [4]:
movies.shape

(979, 6)


🧐 The dataframe has 979 rows and 6 columns or Series. 


In [5]:
type(False)

bool


☝🏻 A gentle reminder that 'True' and 'False' are boolean values as seen when applied the type method. 



⭐️ We want now to filter within the 'duration' Series the observations that have a minimum of 200 mins lenght such as 'The Godfather: Part II'.



⭐️ Here we are going to create a python list of boolean's values to be 'True' if the duration of a movie is at least 200 mins and 'False' if it falls under. 



🧐 In other words, we want to create a list of booleans telling us which rows match a given condition. 


In [12]:
booleans = []
for length in movies.duration:
    if length >= 200:
        booleans.append(True)
    else:
        booleans.append(False)


☝🏻 A list just like a Series is iterable so we can apply a for loop to go through all observations contained within one Series.


In [13]:
booleans[0:5]

[False, False, True, False, False]


✅ Our first test appears to show that we're advancing rightly towards an efficient way to check if the first 5 rows match our conditions by returning a boolean value.  


In [16]:
len(booleans)

979


✅ The lenght of our variable has the same number of rows contained in our dataframe. 




⭐️ Let's now convert our python list into a dataframe Series. 


In [18]:
is_long = pd.Series(booleans)


⭐️ The 'Series()' method converts the booleans into a Series. 


In [19]:
is_long.head()

0    False
1    False
2     True
3    False
4    False
dtype: bool

In [20]:
type(is_long)

pandas.core.series.Series


✅ Tests are successfully run. Our booleans have been converted into a Series. 


In [21]:
movies[is_long]

star_rating                                          title  \
2            9.1                         The Godfather: Part II   
7            8.9  The Lord of the Rings: The Return of the King   
17           8.7                                  Seven Samurai   
78           8.4                    Once Upon a Time in America   
85           8.4                             Lawrence of Arabia   
142          8.3              Lagaan: Once Upon a Time in India   
157          8.2                             Gone with the Wind   
204          8.1                                        Ben-Hur   
445          7.9                           The Ten Commandments   
476          7.8                                         Hamlet   
630          7.7                                      Malcolm X   
767          7.6                It's a Mad, Mad, Mad, Mad World   

    content_rating      genre  duration  \
2                R      Crime       200   
7            PG-13  Adventure       201   
17         UNRATED      Drama       207   
78               R      Crime       229   
85              PG  Adventure       216   
142             PG  Adventure       224   
157              G      Drama       238   
204              G  Adventure       212   
445       APPROVED  Adventure       220   
476          PG-13      Drama       242   
630          PG-13  Biography       202   
767       APPROVED     Action       205   

                                           actors_list  
2    [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
7    [u'Elijah Wood', u'Viggo Mortensen', u'Ian McK...  
17   [u'Toshir\xf4 Mifune', u'Takashi Shimura', u'K...  
78   [u'Robert De Niro', u'James Woods', u'Elizabet...  
85   [u"Peter O'Toole", u'Alec Guinness', u'Anthony...  
142  [u'Aamir Khan', u'Gracy Singh', u'Rachel Shell...  
157  [u'Clark Gable', u'Vivien Leigh', u'Thomas Mit...  
204  [u'Charlton Heston', u'Jack Hawkins', u'Stephe...  
445  [u'Charlton Heston', u'Yul Brynner', u'Anne Ba...  
476  [u'Kenneth Branagh', u'Julie Christie', u'Dere...  
630  [u'Denzel Washington', u'Angela Bassett', u'De...  
767  [u'Spencer Tracy', u'Milton Berle', u'Ethel Me...


✅ We use a bracket annotation to pass a Series of true and false into our dataframe. In doing so, pandas effectively alters the dataframe with all columns showing rows in which the movie's duration was at least 200 minutes.



⭐️ Now that the logic has been understood, we are going to see an easier way to put those ations into a cleaner coding fashion without writting the for loop. 


In [22]:
is_long = movies.duration >= 200
is_long.head() 

0    False
1    False
2     True
3    False
4    False
Name: duration, dtype: bool

In [23]:
movies[is_long]

star_rating                                          title  \
2            9.1                         The Godfather: Part II   
7            8.9  The Lord of the Rings: The Return of the King   
17           8.7                                  Seven Samurai   
78           8.4                    Once Upon a Time in America   
85           8.4                             Lawrence of Arabia   
142          8.3              Lagaan: Once Upon a Time in India   
157          8.2                             Gone with the Wind   
204          8.1                                        Ben-Hur   
445          7.9                           The Ten Commandments   
476          7.8                                         Hamlet   
630          7.7                                      Malcolm X   
767          7.6                It's a Mad, Mad, Mad, Mad World   

    content_rating      genre  duration  \
2                R      Crime       200   
7            PG-13  Adventure       201   
17         UNRATED      Drama       207   
78               R      Crime       229   
85              PG  Adventure       216   
142             PG  Adventure       224   
157              G      Drama       238   
204              G  Adventure       212   
445       APPROVED  Adventure       220   
476          PG-13      Drama       242   
630          PG-13  Biography       202   
767       APPROVED     Action       205   

                                           actors_list  
2    [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
7    [u'Elijah Wood', u'Viggo Mortensen', u'Ian McK...  
17   [u'Toshir\xf4 Mifune', u'Takashi Shimura', u'K...  
78   [u'Robert De Niro', u'James Woods', u'Elizabet...  
85   [u"Peter O'Toole", u'Alec Guinness', u'Anthony...  
142  [u'Aamir Khan', u'Gracy Singh', u'Rachel Shell...  
157  [u'Clark Gable', u'Vivien Leigh', u'Thomas Mit...  
204  [u'Charlton Heston', u'Jack Hawkins', u'Stephe...  
445  [u'Charlton Heston', u'Yul Brynner', u'Anne Ba...  
476  [u'Kenneth Branagh', u'Julie Christie', u'Dere...  
630  [u'Denzel Washington', u'Angela Bassett', u'De...  
767  [u'Spencer Tracy', u'Milton Berle', u'Ethel Me...


🧐 Let's look at another SIMPLER way to achieve the result by writting less code ! 'cause less is better.


In [27]:
movies[movies.duration >= 200]

star_rating                                          title  \
2            9.1                         The Godfather: Part II   
7            8.9  The Lord of the Rings: The Return of the King   
17           8.7                                  Seven Samurai   
78           8.4                    Once Upon a Time in America   
85           8.4                             Lawrence of Arabia   
142          8.3              Lagaan: Once Upon a Time in India   
157          8.2                             Gone with the Wind   
204          8.1                                        Ben-Hur   
445          7.9                           The Ten Commandments   
476          7.8                                         Hamlet   
630          7.7                                      Malcolm X   
767          7.6                It's a Mad, Mad, Mad, Mad World   

    content_rating      genre  duration  \
2                R      Crime       200   
7            PG-13  Adventure       201   
17         UNRATED      Drama       207   
78               R      Crime       229   
85              PG  Adventure       216   
142             PG  Adventure       224   
157              G      Drama       238   
204              G  Adventure       212   
445       APPROVED  Adventure       220   
476          PG-13      Drama       242   
630          PG-13  Biography       202   
767       APPROVED     Action       205   

                                           actors_list  
2    [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
7    [u'Elijah Wood', u'Viggo Mortensen', u'Ian McK...  
17   [u'Toshir\xf4 Mifune', u'Takashi Shimura', u'K...  
78   [u'Robert De Niro', u'James Woods', u'Elizabet...  
85   [u"Peter O'Toole", u'Alec Guinness', u'Anthony...  
142  [u'Aamir Khan', u'Gracy Singh', u'Rachel Shell...  
157  [u'Clark Gable', u'Vivien Leigh', u'Thomas Mit...  
204  [u'Charlton Heston', u'Jack Hawkins', u'Stephe...  
445  [u'Charlton Heston', u'Yul Brynner', u'Anne Ba...  
476  [u'Kenneth Branagh', u'Julie Christie', u'Dere...  
630  [u'Denzel Washington', u'Angela Bassett', u'De...  
767  [u'Spencer Tracy', u'Milton Berle', u'Ethel Me...


✅ This was effectively conducted by writting only one line of code. 


# 🎩 Bonus tips : iPython | Jupyter Notebook ONLY


🤠 Let's assume you want to study the 'genre' of movies fitting our condition of a minimum of 200 mins length. 

Let's see what we can use here :


In [28]:
movies[movies.duration >= 200].genre

2          Crime
7      Adventure
17         Drama
78         Crime
85     Adventure
142    Adventure
157        Drama
204    Adventure
445    Adventure
476        Drama
630    Biography
767       Action
Name: genre, dtype: object


➕ Or using bracket annotation : 


In [30]:
movies[movies.duration >= 200]['genre']

2          Crime
7      Adventure
17         Drama
78         Crime
85     Adventure
142    Adventure
157        Drama
204    Adventure
445    Adventure
476        Drama
630    Biography
767       Action
Name: genre, dtype: object


❗️Best practices tip : A good pandas practicioner would favor the use of 'loc' to select rows and columns by labels. 


In [32]:
movies.loc[movies.duration >= 200, 'genre']

2          Crime
7      Adventure
17         Drama
78         Crime
85     Adventure
142    Adventure
157        Drama
204    Adventure
445    Adventure
476        Drama
630    Biography
767       Action
Name: genre, dtype: object



🙏🏻 Thank you !

👋🏻 See you in the next one !
